### PR TITLE
feat(voice): keyboard shortcuts for toggle and hold-to-talk (#441.2)

### PR DESCRIPTION
**Pull request by jamesbrink**
Closes sub-item 2 of #441.

## Summary

- **Toggle hotkey** (default `Cmd/Ctrl+Shift+M`): starts recording when idle, stops when recording, cancels when starting or transcribing. This combo is free in the existing shortcut scheme.
- **Hold-to-talk** (default `Right Alt / Right Option` on **macOS only**; disabled by default on Windows/Linux to avoid conflicting with AltGr text entry — opt in via Settings → Keyboard): walkie-talkie pattern — hold to record, release to stop and transcribe.

## How rebinding works

Settings → **Keyboard** (new section) shows both hotkeys with **Rebind**, **Reset**, and **Disable** controls. Clicking **Rebind** enters a capture mode; the next keydown is recorded and persisted via \`setAppSetting\`. Changes propagate immediately through Zustand — no app restart needed. **Reset** restores the platform-aware default (which is \`null\`/disabled for hold-to-talk on Windows/Linux).

## Edge cases

- \`event.repeat === true\` is silently ignored so OS key-repeat doesn't multi-fire start.
- Window blur during a hold is treated as a key-release (recording stops cleanly, not stuck on). This now applies to ANY active recording, not just hold-to-talk — Cmd+Tab away during a toggle/button-triggered recording also stops cleanly.
- Blur arriving during the brief \`starting\` state (before \`startVoiceRecording\` resolves) routes through \`cancel()\` and rolls the provider back, so the mic doesn't go hot in the background after the user has switched apps.
- A late \`keyup\` arriving after blur already cleared \`holdActive\` is a no-op.
- AltGr (Right Alt on most non-US keyboard layouts) is rejected at runtime in three forms: \`e.key === "AltGraph"\`, \`e.getModifierState("AltGraph")\`, and the \`Ctrl+Alt+code=AltRight\` signature. Combined with the platform-aware default, this means typing \`@\`/\`{}\`/\`ñ\` on non-US layouts never accidentally triggers voice.
- Hotkeys are gated when an overlay owns input focus (settings, modal, command palette, fuzzy finder) — same gating pattern as \`useKeyboardShortcuts.ts\`. Stop/cancel/release stay ungated so an in-flight recording can always be ended.

## Implementation

- \`utils/voiceHotkeys.ts\` — small dependency-free constants module (\`DEFAULT_TOGGLE_HOTKEY\`, \`DEFAULT_HOLD_HOTKEY_MAC\`, \`isMacPlatform\`, \`getDefaultHoldHotkey\`). Imported by both the hook and the Zustand slice.
- \`hooks/useVoiceHotkey.ts\` — new hook + \`createVoiceHotkeyHandlers\` factory (exported for testing) + format helpers + AltGr detection
- \`hooks/useVoiceHotkey.test.ts\` — 26 tests covering all state machine transitions, AltGr signatures, the overlay gate, and post-Escape preservation
- \`hooks/useVoiceInput.ts\` — central blur listener stops any active recording (mic button, toggle hotkey, hold-to-talk); cancellation rollback after \`startVoiceRecording\` await
- \`stores/slices/settingsSlice.ts\` — \`voiceToggleHotkey\` / \`voiceHoldHotkey\` Zustand state with platform-aware defaults
- \`App.tsx\` — loads persisted hotkey prefs on startup
- \`components/settings/sections/KeyboardSettings.tsx\` — settings UI with rebind/reset/disable; hotkey badge composes \`.iconBtn\` so it visually matches the adjacent buttons exactly
- \`components/chat/ChatInputArea.tsx\` — wires \`useVoiceHotkey(voice, toggleHotkey, holdHotkey)\` after the voice controller is created

## Test plan

- [x] \`bun run test\` passes (1181 tests)
- [x] \`bunx tsc -b\` clean
- [x] \`bun run lint:css\` passes
- [ ] Toggle \`Cmd+Shift+M\` starts/stops voice in the chat composer
- [ ] Hold \`Right Option\` (macOS) records; release transcribes
- [ ] Cmd+Tab during hold OR during a toggle-triggered recording stops cleanly
- [ ] Settings → Keyboard section visible; Rebind captures new key; Reset restores platform default; Disable shows "—"

### CODE REVIEW

File: src/ui/src/hooks/useVoiceHotkey.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+} {
+  let holdActive = false;
+
+  return {
+    onKeyDown(e: KeyboardEvent) {
+      // Suppress repeated hold-key events so the OS key-repeat doesn't
+      // re-trigger start(). The !e.repeat guard on the hold branch below
+      // already handles this, but we also preventDefault so the webview
+      // doesn't receive spurious Alt/Option characters while holding.
+      if (e.repeat) {
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/hooks/useVoiceHotkey.ts]: Addressed in 88a54fb. The repeat branch now also calls `preventDefault()` when the event matches `toggleHotkey`, so a no-modifier rebind (e.g. plain `a`) no longer leaks repeated characters into focused inputs while held.

File: src/ui/src/components/settings/sections/KeyboardSettings.tsx [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        setRebinding(null);
+        return;
+      }
+
+      if (rebinding === "toggle") {
+        const combo = captureToggleCombo(e);
+        if (combo) void persist("voice_toggle_hotkey", combo, setVoiceToggleHotkey);
+      } else {
+        void persist("voice_hold_hotkey", e.code, setVoiceHoldHotkey);
+      }
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/settings/sections/KeyboardSettings.tsx:112]: Addressed in 88a54fb. Added a closure-local `captured` flag in the rebind effect that flips synchronously on the first capturing keydown, so subsequent keys arriving before the React state/listener teardown are ignored (and only one `persist()` runs).

File: src/ui/src/components/settings/sections/KeyboardSettings.tsx [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            {t("keyboard_voice_toggle_desc")}
+          </div>
+        </div>
+        <div className={styles.settingControl}>
+          <div className={styles.inlineControl}>
+            <code style={{ fontFamily: "var(--font-mono)", fontSize: 13, padding: "4px 8px", background: "var(--selected-bg)", border: "1px solid var(--divider)", borderRadius: 4, minWidth: 60, textAlign: "center" }}>
+              {rebinding === "toggle"
+                ? t("keyboard_press_key")
+                : formatToggleHotkey(voiceToggleHotkey, isMac)}
+            </code>
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/settings/sections/KeyboardSettings.tsx:142]: Addressed in 88a54fb. Extracted the inline style into a `.hotkeyBadge` class in `Settings.module.css` and applied it to both `<code>` elements.

File: src/ui/src/components/settings/sections/KeyboardSettings.tsx [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+function captureToggleCombo(e: KeyboardEvent): string | null {
+  const modifierKeys = new Set(["Meta", "Control", "Shift", "Alt"]);
+  if (modifierKeys.has(e.key)) return null;
+  const parts: string[] = [];
+  if (e.metaKey || e.ctrlKey) parts.push("mod");
+  if (e.shiftKey) parts.push("shift");
+  if (e.altKey) parts.push("alt");
+  parts.push(e.key.toLowerCase());
+  return parts.join("+");
+}
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/settings/sections/KeyboardSettings.tsx:31]: Addressed in 88a54fb. Added a `normalizeToggleKey()` helper that maps `+` to the stable token `plus`. Both `captureToggleCombo` (in KeyboardSettings) and `matchesToggle` (in useVoiceHotkey) now route the key through it, so storage stays unambiguous (`mod+shift+plus`) and matching still resolves the literal `+` keypress. `formatToggleHotkey` displays it as `+`.

File: src/ui/src/stores/slices/settingsSlice.ts [hunk 5]
```diff
@@ -115,4 +119,8 @@ export const createSettingsSlice: StateCreator<
   setDisable1mContext: (v) => set({ disable1mContext: v }),
   editorGitGutterBase: "head",
   setEditorGitGutterBase: (value) => set({ editorGitGutterBase: value }),
+  voiceToggleHotkey: "mod+shift+m",
+  setVoiceToggleHotkey: (hotkey) => set({ voiceToggleHotkey: hotkey }),
+  voiceHoldHotkey: "AltRight",
+  setVoiceHoldHotkey: (hotkey) => set({ voiceHoldHotkey: hotkey }),
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/stores/slices/settingsSlice.ts]: Addressed in 5e36dfc. The hold default is now platform-aware via `getDefaultHoldHotkey()` in `src/utils/voiceHotkeys.ts` — `AltRight` on macOS, `null` elsewhere. Both the Zustand slice initial value and the Reset button consult the same helper, so non-Mac users get a disabled-by-default hold key and have to opt in via Settings → Keyboard.

File: src/ui/src/hooks/useVoiceHotkey.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+          v.stop();
+        } else if (v.state === "starting" || v.state === "transcribing") {
+          v.cancel();
+        } else if (v.state === "idle") {
+          void v.start();
+        }
+        return;
+      }
+
+      if (holdHotkey && e.code === holdHotkey && !holdActive && v.state === "idle") {
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/hooks/useVoiceHotkey.ts]: Addressed in 5e36dfc. `useVoiceHotkey` now passes an `isInputBlocked()` callback into the handler factory that consults the same store flags as `useKeyboardShortcuts.ts` (`settingsOpen || activeModal || commandPaletteOpen || fuzzyFinderOpen`). START actions short-circuit when an overlay owns input focus; stop/cancel/release stay ungated so an in-flight recording can always be ended. Two new tests cover the gate (blocks start, allows stop).

File: src/ui/src/components/settings/sections/KeyboardSettings.tsx [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+    const onKeyDown = (e: KeyboardEvent) => {
+      e.preventDefault();
+      e.stopPropagation();
+      if (captured) return;
+
+      if (e.key === "Escape") {
+        captured = true;
+        setRebinding(null);
+        return;
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/settings/sections/KeyboardSettings.tsx:101]: Addressed in 5e36dfc. Reordered the rebind `onKeyDown` so the `captured` check runs BEFORE `preventDefault()` / `stopPropagation()`. Post-Escape keystrokes that arrive in the brief window before the effect cleanup are now passed through cleanly instead of being swallowed.

File: src/ui/src/hooks/useVoiceHotkey.ts [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+          void v.start();
+        }
+        return;
+      }
+
+      if (
+        holdHotkey &&
+        e.code === holdHotkey &&
+        !holdActive &&
+        v.state !== "recording" &&
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/hooks/useVoiceHotkey.ts:148]: Addressed in 5e36dfc. Two layers: (1) the default hold key is now platform-aware (`AltRight` on macOS only, `null` elsewhere) — see also the reply to the related finding on `settingsSlice.ts`. (2) Runtime `isAltGr()` helper rejects AltGr presses regardless of binding — checks `e.key === "AltGraph"`, `e.getModifierState("AltGraph")`, and the `Ctrl+Alt+code=AltRight` signature. So even a Mac user with a German layout can keep the default and AltGr typing still works. Three new tests cover all three AltGr signatures.

File: src/ui/src/stores/slices/settingsSlice.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/stores/slices/settingsSlice.ts]: Addressed in 5e36dfc. Created `src/utils/voiceHotkeys.ts` — a small dependency-free module that owns `DEFAULT_TOGGLE_HOTKEY`, `DEFAULT_HOLD_HOTKEY_MAC`, `isMacPlatform()`, and `getDefaultHoldHotkey()`. Both `useVoiceHotkey.ts` (which re-exports for backward compat) and `settingsSlice.ts` import from there. As a bonus this resolved a circular import that surfaced when the hook started importing `useAppStore` for overlay gating.

File: src/ui/src/components/settings/sections/KeyboardSettings.tsx [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+type RebindTarget = "toggle" | "hold" | null;
+
+/** Capture a toggle-hotkey combo from a keydown event. Returns null for modifier-only presses. */
+function captureToggleCombo(e: KeyboardEvent): string | null {
+  const modifierKeys = new Set(["Meta", "Control", "Shift", "Alt"]);
+  if (modifierKeys.has(e.key)) return null;
+  const parts: string[] = [];
+  if (e.metaKey || e.ctrlKey) parts.push("mod");
+  if (e.shiftKey) parts.push("shift");
+  if (e.altKey) parts.push("alt");
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/settings/sections/KeyboardSettings.tsx:27]: Addressed in 5e36dfc. `captureToggleCombo` now includes `"AltGraph"` in the modifier-only set, so the rebind dialog skips that initial keydown and waits for the intended non-modifier key. Together with the runtime `isAltGr()` rejection in the hook, AltGr presses are filtered consistently in both the rebind and runtime paths.

File: src/ui/src/hooks/useVoiceInput.ts [hunk 10]
```diff
@@ -332,6 +332,16 @@ export function useVoiceInput(
...
 
+  // Stop any active recording when the window loses focus, regardless of how
+  // it was started (mic button, toggle hotkey, hold-to-talk). Without this,
+  // a Cmd+Tab away from the app would leave the mic hot in the background.
+  // stop() is internally gated — it's a no-op when no recording is in flight.
+  useEffect(() => {
+    const onBlur = () => stop();
+    window.addEventListener("blur", onBlur);
+    return () => window.removeEventListener("blur", onBlur);
+  }, [stop]);
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/hooks/useVoiceInput.ts]: Addressed in 9d6b02e. Two-part fix for the blur-during-`starting` race: (1) the blur handler in `useVoiceInput` now routes to `cancel()` when `stateRef.current === "starting"` (since `stop()` is a no-op before `nativeProviderRef.current` is set). (2) `start()` checks `cancelledRef.current` immediately AFTER `await startVoiceRecording(...)` resolves — if cancel fired during the await, it calls `cancelVoiceRecording()` to roll back the provider and returns without becoming active. New test `"rolls back the native start when cancelled mid-await"` reproduces the original race and verifies `cancelVoiceRecording` is called.

File: src/ui/src/components/settings/sections/KeyboardSettings.tsx [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+  getDefaultHoldHotkey,
+  normalizeToggleKey,
+} from "../../../hooks/useVoiceHotkey";
+import styles from "../Settings.module.css";
+
+function isMacPlatform(): boolean {
+  if (typeof navigator === "undefined") return false;
+  return /Mac/.test(navigator.platform) || /Mac OS X/.test(navigator.userAgent);
+}
+
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/settings/sections/KeyboardSettings.tsx]: Addressed in 9d6b02e. Removed the local `isMacPlatform()` from `KeyboardSettings.tsx` and now import it from `src/utils/voiceHotkeys.ts` — same source `getDefaultHoldHotkey()` uses internally. One definition, no drift risk.

File: src/ui/src/hooks/useVoiceHotkey.ts [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    ControlLeft: { mac: "Left ⌃", other: "Left Ctrl" },
+    ShiftRight: { mac: "Right ⇧", other: "Right Shift" },
+    ShiftLeft: { mac: "Left ⇧", other: "Left Shift" },
+    MetaRight: { mac: "Right ⌘", other: "Right Meta" },
+    MetaLeft: { mac: "Left ⌘", other: "Left Meta" },
+    Space: { mac: "Space", other: "Space" },
+    F13: { mac: "F13", other: "F13" },
+    F14: { mac: "F14", other: "F14" },
+    F15: { mac: "F15", other: "F15" },
+  };
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/hooks/useVoiceHotkey.ts]: Addressed in 9d6b02e. Hoisted the lookup table to a module-scope `HOLD_KEY_DISPLAY` const so it is allocated once at module load instead of on every `formatHoldHotkey()` call.

File: src/ui/src/stores/slices/settingsSlice.ts [hunk 13]
```diff
@@ -115,4 +123,8 @@ export const createSettingsSlice: StateCreator<
   setDisable1mContext: (v) => set({ disable1mContext: v }),
   editorGitGutterBase: "head",
   setEditorGitGutterBase: (value) => set({ editorGitGutterBase: value }),
+  voiceToggleHotkey: DEFAULT_TOGGLE_HOTKEY,
+  setVoiceToggleHotkey: (hotkey) => set({ voiceToggleHotkey: hotkey }),
+  voiceHoldHotkey: getDefaultHoldHotkey(),
+  setVoiceHoldHotkey: (hotkey) => set({ voiceHoldHotkey: hotkey }),
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/stores/slices/settingsSlice.ts:129]: Addressed in 9d6b02e (PR description update). The PR body now explicitly states the hold default is `Right Alt/Right Option` on **macOS only**, and `null` (disabled) on Windows/Linux to avoid AltGr text-entry conflicts. The Test plan and Edge cases sections were also updated to reflect the platform-aware default and the Reset behavior.
